File: src/statusline.test.ts
import { describe, expect, test } from 'bun:test';
import {
  boardKeyForCwd,
  DEFAULT_STATUSLINE_TEMPLATE,
  renderStatusline,
  type StatuslineData,
  truncateTitle,
} from './statusline';

const empty: StatuslineData = { inbox: 0, stale: 0, doing: 0 };

const withMine: StatuslineData = {
  mine: { ref: 'rocky-todo-12', title: 'statusline API 추가', comments: 0 },
  inbox: 0,
  stale: 0,
  doing: 1,
};

describe('renderStatusline — 그룹', () => {
  test('보여줄 게 없으면 빈 문자열이다 (호출부가 아무것도 출력하지 않게)', () => {
    expect(renderStatusline(DEFAULT_STATUSLINE_TEMPLATE, empty)).toBe('');
  });

  test('0 인 숫자는 자기 그룹을 통째로 지운다', () => {
    expect(renderStatusline('[⏺{doing}][ ✉{inbox}]', { ...empty, doing: 3 })).toBe('⏺3');
  });

  test('앞 그룹이 사라져 남은 공백은 trim 된다', () => {
    expect(renderStatusline('[⏺ {mine.ref}][  ✉{inbox}]', { ...empty, inbox: 2 })).toBe('✉2');
  });

  test('그룹 안 placeholder 가 하나라도 차 있으면 그룹이 남는다', () => {
    expect(renderStatusline('[{mine.ref} {mine.title}]', withMine)).toBe(
      'rocky-todo-12 statusline API 추가',
    );
  });

  test('placeholder 가 없는 그룹은 순수 장식이라 늘 남는다', () => {
    expect(renderStatusline('[보드]', empty)).toBe('보드');
  });

  test('모르는 placeholder 는 그대로 남는다 — 오타가 눈에 보여야 한다', () => {
    expect(renderStatusline('[{mine.titel}]', withMine)).toBe('{mine.titel}');
  });

  test('닫히지 않은 대괄호는 리터럴로 되돌린다 — 조용히 잘려나가지 않게', () => {
    expect(renderStatusline('[⏺{doing}', { ...empty, doing: 2 })).toBe('[⏺2');
  });
});

describe('renderStatusline — ANSI', () => {
  // 색을 별도 DSL 로 만들지 않고 템플릿에 이스케이프를 직접 적게 한 선택의 유일한 위험이
  // 여기다. `ESC[33m` 의 `[` 를 그룹 시작으로 읽으면 색을 넣은 템플릿이 통째로 깨진다.
  const yellow = '\u001b[33m';
  const reset = '\u001b[0m';

  test('ESC 뒤의 `[` 는 그룹이 아니라 리터럴이다', () => {
    expect(renderStatusline(`[${yellow}⏺{doing}${reset}]`, { ...empty, doing: 4 })).toBe(
      `${yellow}⏺4${reset}`,
    );
  });

  test('색을 입힌 그룹도 값이 비면 이스케이프까지 같이 사라진다', () => {
    expect(renderStatusline(`[${yellow}⏺{doing}${reset}][ ✉{inbox}]`, { ...empty, inbox: 1 })).toBe(
      '✉1',
    );
  });
});

describe('renderStatusline — mine', () => {
  test('기본 템플릿은 이 세션의 항목을 제목까지 싣는다', () => {
    expect(renderStatusline(DEFAULT_STATUSLINE_TEMPLATE, withMine)).toBe(
      '⏺ rocky-todo-12 statusline API 추가',
    );
  });

  test('댓글이 달리면 세그먼트가 생긴다 — 사람이 남긴 말이 곧 이 기능의 목적이다', () => {
    const line = renderStatusline(DEFAULT_STATUSLINE_TEMPLATE, {
      ...withMine,
      mine: { ...withMine.mine!, comments: 3 },
    });
    expect(line).toBe('⏺ rocky-todo-12 statusline API 추가 💬3');
  });

  test('mine 이 없으면 mine 관련 그룹은 전부 사라진다', () => {
    expect(renderStatusline(DEFAULT_STATUSLINE_TEMPLATE, { ...empty, stale: 2 })).toBe('⚠2');
  });
});

describe('truncateTitle', () => {
  test('짧은 제목은 그대로 둔다', () => {
    expect(truncateTitle('짧다')).toBe('짧다');
  });

  test('긴 제목은 잘라내고 잘렸음을 표시한다', () => {
    expect(truncateTitle('가'.repeat(40))).toBe(`${'가'.repeat(30)}…`);
  });

  test('코드포인트 단위로 자른다 — 이모지가 반 토막 나지 않게', () => {
    expect(truncateTitle('🙂'.repeat(5), 3)).toBe('🙂🙂🙂…');
  });
});

describe('boardKeyForCwd', () => {
  const boards = [
    { key: 'rocky-todo', path: '/Users/x/dev/rocky-todo' },
    { key: 'rocky' },
    { key: 'ogpeek', path: '/Users/x/dev/ogpeek' },
  ];

  test('cwd 가 없으면 아무 보드도 고르지 않는다', () => {
    expect(boardKeyForCwd(boards, undefined)).toBeUndefined();
  });

  test('boards.path 아래면 그 보드다', () => {
    expect(boardKeyForCwd(boards, '/Users/x/dev/rocky-todo')).toBe('rocky-todo');
  });

  test('path 비교는 경로 경계까지 본다 — `/a/b` 가 `/a/bc` 에 걸리면 안 된다', () => {
    expect(boardKeyForCwd([{ key: 'og', path: '/Users/x/dev/og' }], '/Users/x/dev/ogpeek')).toBe(
      undefined,
    );
  });

  test('path 가 없어도 보드 key 가 경로 세그먼트면 잡는다', () => {
    expect(boardKeyForCwd(boards, '/Users/x/orca/rocky/eelpout')).toBe('rocky');
  });

  test('워크트리 — basename 이 아니라 세그먼트를 봐야 원본 보드가 잡힌다', () => {
    const cwd = '/Users/x/dev/rocky-todo/.claude/worktrees/todo-12';
    expect(boardKeyForCwd(boards, cwd)).toBe('rocky-todo');
  });

  test('세그먼트 후보가 여럿이면 더 구체적인(긴) key 를 고른다', () => {
    const cwd = '/Users/x/rocky/rocky-todo/sub';
    expect(boardKeyForCwd([{ key: 'rocky' }, { key: 'rocky-todo' }], cwd)).toBe('rocky-todo');
  });
});

File: crates/rocky-todo-core/tests/statusline_test.rs
//! TS 원본 `src/statusline.test.ts` 포팅.

use rocky_todo_core::statusline::{
    board_key_for_cwd, render_statusline, truncate_title, BoardLocation, StatuslineData,
    StatuslineMine, DEFAULT_STATUSLINE_TEMPLATE, STATUSLINE_TITLE_MAX,
};

fn empty() -> StatuslineData {
    StatuslineData::default()
}

fn with_mine() -> StatuslineData {
    StatuslineData {
        mine: Some(StatuslineMine {
            r#ref: "rocky-todo-12".into(),
            title: "statusline API 추가".into(),
            comments: 0,
        }),
        inbox: 0,
        stale: 0,
        doing: 1,
    }
}

fn render(template: &str, data: &StatuslineData) -> String {
    render_statusline(template, data, STATUSLINE_TITLE_MAX)
}

#[test]
fn nothing_to_show_renders_empty() {
    assert_eq!(render(DEFAULT_STATUSLINE_TEMPLATE, &empty()), "");
}

#[test]
fn zero_counts_erase_their_group() {
    assert_eq!(
        render(
            "[⏺{doing}][ ✉{inbox}]",
            &StatuslineData {
                doing: 3,
                ..empty()
            }
        ),
        "⏺3"
    );
}

#[test]
fn leading_whitespace_from_erased_groups_is_trimmed() {
    assert_eq!(
        render(
            "[⏺ {mine.ref}][  ✉{inbox}]",
            &StatuslineData {
                inbox: 2,
                ..empty()
            }
        ),
        "✉2"
    );
}

#[test]
fn group_survives_when_any_placeholder_filled() {
    assert_eq!(
        render("[{mine.ref} {mine.title}]", &with_mine()),
        "rocky-todo-12 statusline API 추가"
    );
}

#[test]
fn decoration_only_group_always_stays() {
    assert_eq!(render("[보드]", &empty()), "보드");
}

#[test]
fn unknown_placeholder_stays_visible() {
    assert_eq!(render("[{mine.titel}]", &with_mine()), "{mine.titel}");
}

#[test]
fn unclosed_bracket_reverts_to_literal() {
    assert_eq!(
        render(
            "[⏺{doing}",
            &StatuslineData {
                doing: 2,
                ..empty()
            }
        ),
        "[⏺2"
    );
}

#[test]
fn esc_bracket_is_literal_not_group() {
    let yellow = "\u{001b}[33m";
    let reset = "\u{001b}[0m";
    assert_eq!(
        render(
            &format!("[{yellow}⏺{{doing}}{reset}]"),
            &StatuslineData {
                doing: 4,
                ..empty()
            }
        ),
        format!("{yellow}⏺4{reset}")
    );
}

#[test]
fn colored_group_disappears_with_escapes_when_empty() {
    let yellow = "\u{001b}[33m";
    let reset = "\u{001b}[0m";
    assert_eq!(
        render(
            &format!("[{yellow}⏺{{doing}}{reset}][ ✉{{inbox}}]"),
            &StatuslineData {
                inbox: 1,
                ..empty()
            }
        ),
        "✉1"
    );
}

#[test]
fn default_template_carries_mine_with_title() {
    assert_eq!(
        render(DEFAULT_STATUSLINE_TEMPLATE, &with_mine()),
        "⏺ rocky-todo-12 statusline API 추가"
    );
}

#[test]
fn comment_segment_appears_when_comments_arrive() {
    let mut data = with_mine();
    data.mine.as_mut().unwrap().comments = 3;
    assert_eq!(
        render(DEFAULT_STATUSLINE_TEMPLATE, &data),
        "⏺ rocky-todo-12 statusline API 추가 💬3"
    );
}

#[test]
fn without_mine_all_mine_groups_disappear() {
    assert_eq!(
        render(
            DEFAULT_STATUSLINE_TEMPLATE,
            &StatuslineData {
                stale: 2,
                ..empty()
            }
        ),
        "⚠2"
    );
}

#[test]
fn truncate_short_title_unchanged() {
    assert_eq!(truncate_title("짧다", STATUSLINE_TITLE_MAX), "짧다");
}

#[test]
fn truncate_long_title_with_ellipsis() {
    let long = "가".repeat(40);
    assert_eq!(
        truncate_title(&long, STATUSLINE_TITLE_MAX),
        format!("{}…", "가".repeat(30))
    );
}

#[test]
fn truncate_cuts_at_code_points() {
    assert_eq!(truncate_title(&"🙂".repeat(5), 3), "🙂🙂🙂…");
}

// ── boardKeyForCwd ──

fn boards() -> Vec<BoardLocation> {
    vec![
        BoardLocation {
            key: "rocky-todo".into(),
            path: Some("/Users/x/dev/rocky-todo".into()),
        },
        BoardLocation {
            key: "rocky".into(),
            path: None,
        },
        BoardLocation {
            key: "ogpeek".into(),
            path: Some("/Users/x/dev/ogpeek".into()),
        },
    ]
}

#[test]
fn no_cwd_picks_nothing() {
    assert_eq!(board_key_for_cwd(&boards(), None), None);
}

#[test]
fn under_board_path_wins() {
    assert_eq!(
        board_key_for_cwd(&boards(), Some("/Users/x/dev/rocky-todo")).as_deref(),
        Some("rocky-todo")
    );
}

#[test]
fn path_comparison_respects_boundaries() {
    let og = vec![BoardLocation {
        key: "og".into(),
        path: Some("/Users/x/dev/og".into()),
    }];
    assert_eq!(board_key_for_cwd(&og, Some("/Users/x/dev/ogpeek")), None);
}

#[test]
fn trailing_slash_is_normalized() {
    let slashed = vec![BoardLocation {
        key: "proj".into(),
        path: Some("/Users/x/dev/proj/".into()),
    }];
    assert_eq!(
        board_key_for_cwd(&slashed, Some("/Users/x/dev/proj")).as_deref(),
        Some("proj")
    );
    assert_eq!(
        board_key_for_cwd(&slashed, Some("/Users/x/dev/proj/sub")).as_deref(),
        Some("proj")
    );
}

#[test]
fn trailing_slash_does_not_skew_length_comparison() {
    let both = vec![
        BoardLocation {
            key: "outer".into(),
            path: Some("/Users/x/dev/".into()),
        },
        BoardLocation {
            key: "inner".into(),
            path: Some("/Users/x/dev/proj".into()),
        },
    ];
    assert_eq!(
        board_key_for_cwd(&both, Some("/Users/x/dev/proj/sub")).as_deref(),
        Some("inner")
    );
}

#[test]
fn key_as_path_segment_matches_without_path() {
    assert_eq!(
        board_key_for_cwd(&boards(), Some("/Users/x/orca/rocky/eelpout")).as_deref(),
        Some("rocky")
    );
}

#[test]
fn worktree_matches_by_segment_not_basename() {
    let cwd = "/Users/x/dev/rocky-todo/.claude/worktrees/todo-12";
    assert_eq!(
        board_key_for_cwd(&boards(), Some(cwd)).as_deref(),
        Some("rocky-todo")
    );
}

#[test]
fn multiple_segment_candidates_pick_longest_key() {
    let bs = vec![
        BoardLocation {
            key: "rocky".into(),
            path: None,
        },
        BoardLocation {
            key: "rocky-todo".into(),
            path: None,
        },
    ];
    assert_eq!(
        board_key_for_cwd(&bs, Some("/Users/x/rocky/rocky-todo/sub")).as_deref(),
        Some("rocky-todo")
    );
}
